src/PiWeb.Volume/Block/Quantization.cs

```diff
@@ -8,136 +8,151 @@
+	}
 
-				result[ i ] = value;
-			}
+	private static bool TryGetDouble( this IReadOnlyDictionary<string, string> options, string name, out double value )
+	{
+		value = default;
+		return
+			options.TryGetValue( name, out var valueString ) &&
+			double.TryParse( valueString, NumberStyles.Float, CultureInfo.InvariantCulture, out value );
+	}
 
-			return true;
-		}
+	private static double[] Calculate( double quality = 75, double quantizationBase = 12, double quantizationGain = 1 )
+	{
+		var scale = QualityScaling( (int)quality );
+		var values = new double[ BlockVolume.N3 ];
 
-		private static bool TryGetDouble( this IReadOnlyDictionary<string, string> options, string name, out double value )
-		{
-			value = default;
-			return
-				options.TryGetValue( name, out var valueString ) &&
-				double.TryParse( valueString, NumberStyles.Float, CultureInfo.InvariantCulture, out value );
-		}
+		var i = 0;
 
-		private static double[] Calculate( double quality = 75, double quantizationBase = 12, double quantizationGain = 1 )
+		for( var z = 0; z < BlockVolume.N; z++ )
+		for( var y = 0; y < BlockVolume.N; y++ )
+		for( var x = 0; x < BlockVolume.N; x++ )
 		{
-			var scale = QualityScaling( (int)quality );
-			var values = new double[ BlockVolume.N3 ];
-
-			var i = 0;
+			var distance = Math.Max( 0, Math.Max( x, Math.Max( y, z ) ) - 1 );
+			var baseValue = quantizationBase * ( 1 + quantizationGain * distance );
 
-			for( var z = 0; z < BlockVolume.N; z++ )
-			for( var y = 0; y < BlockVolume.N; y++ )
-			for( var x = 0; x < BlockVolume.N; x++ )
-			{
-				var distance = Math.Max( 0, Math.Max( x, Math.Max( y, z ) ) - 1 );
-				var baseValue = quantizationBase * ( 1 + quantizationGain * distance );
+			var value = Math.Max( 1, Math.Min( short.MaxValue, ( baseValue * 2.0 * scale + 50 ) / 100 ) );
+			values[ i++ ] = 1.0 / value;
+		}
 
-				var value = Math.Max( 1, Math.Min( short.MaxValue, ( baseValue * 2.0 * scale + 50 ) / 100 ) );
-				values[ i++ ] = 1.0 / value;
-			}
+		return values;
+	}
 
-			return values;
-		}
+	/// <summary>
+	/// Quality scaling as in libjpeg.
+	/// </summary>
+	private static int QualityScaling( int quality )
+	{
+		if( quality <= 0 ) quality = 1;
+		if( quality > 100 ) quality = 100;
 
-		/// <summary>
-		/// Quality scaling as in libjpeg.
-		/// </summary>
-		private static int QualityScaling( int quality )
-		{
-			if( quality <= 0 ) quality = 1;
-			if( quality > 100 ) quality = 100;
+		if( quality < 50 )
+			quality = 5000 / quality;
+		else
+			quality = 200 - quality * 2;
 
-			if( quality < 50 )
-				quality = 5000 / quality;
-			else
-				quality = 200 - quality * 2;
+		return quality;
+	}
 
-			return quality;
+	/// <summary>
+	/// Multiplies the <paramref name="quantization"/> with <paramref name="values"/> and stores the result in
+	/// <paramref name="result"/>.
+	/// </summary>
+	public static void Apply( ReadOnlySpan<double> quantization, ReadOnlySpan<double> values, Span<double> result )
+	{
```

Comment:
You can also "cast" the spans to Vector512 spans. That way you can iterate through the Vector512 values easily without slicing and you can also assign the result directly without the CopyTo call.